src/puppetlabs/puppetdb/query_eng/engine.clj

```diff
@@ -2516,28 +2667,286 @@
                    " "
                    (trs "Check your query and try again."))))))
 
+(defn simple-coll? [x] (and (coll? x) (not (map? x))))
+
+(defn extract-deps [m plan]
```

Comment:
Looks like extract-deps can't find a match for inventory queries that use the `.match("pattern.*")` syntax available in the inventory endpoint. The query doesn't cause an error in this case, drop-joins throws when we it can't find a match and then the optimization isn't applied and the error is logged. You can trigger this case with the following query: 

`["from", "inventory",
           ["extract", ["facts"], ["=", "facts.match(\"trust.*\").certname", "pg1.vm"]]]` 

The pattern it can't match in extract-deps is: 
`#puppetlabs.puppetdb.query_eng.engine.JsonContainsExpression{:field "trusted", :column-data {:type :queryable-json, :projectable-json? true, :queryable? true, :field #sql/raw "(fs.stable||fs.volatile)->'trusted'", :field-type :raw, :join-deps #{:fs}}, :value "?", :array-in-path false}`